### ISSUE TITLE
Issues with focus/blur when using flex layout

### ISSUE THREAD

**Issue by carpii** [rivo/tview#870]
I'm using a flex layout, containing two frames. Top frame contains a form, bottom frame contains a table

I'd like to handle focus/blur on both frames, but the docs suggest this is not possible when they are contained in a flex or grid.

So instead, I attach handlers to the form and table, but these never fire either. It's not clear whether this is a bug, or just a known issue. 

It feels like quite a major limitation though, since all but the most trivial apps will require some sort of layout container  

	package ui

	import (
	"github.com/gdamore/tcell/v2"
	"github.com/rivo/tview"
	)

	func ui_debug(s string) {
	// noop
	}

	func UI_test_focus() error {
	app := tview.NewApplication().EnableMouse(true)

	// create test form
	form := tview.NewForm().
		AddInputField("Test", "TEST INPUT 1", 0, nil, nil).
		AddInputField("Test", "TEST INPUT 2", 0, nil, nil)

	// create test table
	color := tcell.ColorWhite
	table := tview.NewTable().SetBorders(false).SetSelectable(true, false)
	ui_table_header_cell(table, 0, 0, tview.AlignLeft, "TEST TABLE")
	table.SetCell(1, 0, tview.NewTableCell("TEST ROW").SetTextColor(color).SetAlign(tview.AlignLeft))
	table.Select(1, 0).SetFixed(1, 0)

	// create top frame
	frame_top := tview.NewFrame(form)
	frame_top.SetBorder(true).SetTitleAlign(tview.AlignLeft).SetTitle(" Frame Top (form) ")

	// create bottom frame
	frame_bottom := tview.NewFrame(table)
	frame_bottom.SetBorder(true).SetTitleAlign(tview.AlignLeft).SetTitle(" Frame Bottom (table) ")

	// attempt to handle focus/blur events - none of these handlers ever fire
	frame_bottom.SetFocusFunc(func() {
		ui_debug("frame_bottom focused")
	}).SetBlurFunc(func() {
		ui_debug("frame_bottom blurred")
	})

	form.SetFocusFunc(func() {
		ui_debug("form focused")
	}).SetBlurFunc(func() {
		ui_debug("form blurred")
	})

	table.SetFocusFunc(func() {
		ui_debug("table focused")
	}).SetBlurFunc(func() {
		ui_debug("table blurred")
	})

	// create flex container
	flex := tview.NewFlex().SetDirection(tview.FlexRow)
	flex.AddItem(frame_top, 0, 3, true)
	flex.AddItem(frame_bottom, 0, 1, false)
	if err := app.SetRoot(flex, true).Run(); err != nil {
		return err
	}

	return nil
	}

**Comment by digitallyserviced** [comment]
I responded to #869... maybe can help?

https://github.com/rivo/tview/issues/869#issuecomment-1694233740

**Comment by rivo** [comment]
Thank you. When I run your code, after clicking on the table first, then on the form, I get this in `debug.log`:

```
[BLUR] frame TOP
[BLUR] form TOP
[FOCUS] table BOTTOM
[BLUR] table BOTTOM
```

And I'm actually surprised that the "blur" event is invoked on the frame and the form. Mostly because [the documentation](https://pkg.go.dev/github.com/rivo/tview#Box.SetBlurFunc) says this:

> SetBlurFunc sets a callback function which is invoked when this primitive loses focus. *This does not apply to container primitives such as Flex or Grid.*

`Frame` and `Form` are both also container primitives. A `Frame` contains another primitive. A `Form` contains multiple other primitives (input fields, checkboxes, buttons etc.). When you click on a flex item to direct the focus to it, the `Flex` component passes the click on to the item itself. Then the item, e.g. your `Form`, passes it on again to the form item at the click position, e.g. an input field. (If there is no item at that position, the last selected item is chosen.)

So your click, and therefore your focus/blur event, is only ever processed by a "leaf" in the layout tree and never by the nodes inbetween that leaf and the root node. (On a side note, when a border is drawn, the [`HasFocus()`](https://pkg.go.dev/github.com/rivo/tview#Primitive) method is called which, for container primitives, will call its contained primitives — essentially traversing the tree.)

Long story short, the behaviour you're seeing is expected. I'm not sure how easy it would be to change this (and how it would affect existing applications). In any case, I would like to understand first what you're trying to achieve. Maybe you can explain a bit why your application needs this? There may be other ways to solve the issue you're having.

**Comment by rivo** [comment]
So it seems that the "blur" events on the frame and the form are called during initialization. They briefly have focus when they are added to the layout and then lose it again when the other elements are added.

**Comment by carpii** [comment]
Ok, thanks. I wanted to make sure its defined behavior I can  rely on, rather than just a bug.

For use case, it will vary. I have lots of these multi frame screens, and was looking for a generic way to detect when focus moves from one frame to another 

Development on this app stalled fort a while, but from what I recall in some cases I need to hit the db the first time the bottom frame is focused, or other times it will change the bottom table from readonly to editable when the frame has focus).

I guess my plan now is to add a generic focus handler to every control, detect the frame it belongs to, and if the frame focus has changed then simulate a frame focus event and run my app specific handler. Unless you have any suggestions for a better way?

Cheers

**Comment by rivo** [comment]
Well, since the same has been requested in #869, I might have to think about how to make this possible in container primitives as well. So far, the implementation is very simple: I keep a reference to the currently focused item. Then I can call `Blur()` on that item when it loses focus.

`SetFocus()` is also very simple: It just updates the reference.

But if I want to include all primitives that the newly (or previously) focused element is contained in, I have to search the entire layout tree for it. There is currently no upwards reference (i.e. a `parent` pointer) on elements and I prefer not to introduce that. (It would bring with it many other problems.)

It's probably not difficult to implement but it will introduce a performance penalty. I guess it's acceptable, though. I don't think anyone has millions of input fields in their application.

**Comment by carpii** [comment]
> There is currently no upwards reference (i.e. a parent pointer) on elements

Ah I remember now. I think that's why I didnt just go ahead and implement the plan I mentioned above (since its not possible to take a control and identify which container it belongs to)

Any improvements you can come up with would be very welcome, and I think would probably be useful for other apps where the UI is dynamically created based on external data.

**Comment by rivo** [comment]
With the latest commit, the "focus" and "blur" notifications are now invoked for all primitives along the hierarchy (see [`SetFocusFunc`](https://github.com/rivo/tview/blob/f9502cf99cec5166dc89e09faab8ef73c18d1e97/box.go#L453) and [`SetBlurFunc`](https://github.com/rivo/tview/blob/f9502cf99cec5166dc89e09faab8ef73c18d1e97/box.go#L470) for details). Please have a look and let me know if this solves your problem.

**Comment by uqix** [comment]
`SetFocusFunc` on form items in flex layout works now, thanks.
